Make only rustc_trans depend on rustc_llvm

hanna-kruppe opened the pull request:
With these changes, only rustc_trans depends directly on rustc_llvm (and no crate gained a new dependency on trans). This means changing LLVM doesn't rebuild librustc or rustc_metadata, only rustc_trans, rustc_driver and the rustc executable
Also, rustc_driver technically doesn't know about LLVM any more (of course, it still handles a ton of options that conceptually refer to LLVM, but it delegates their implementation to trans).

What I *didn't* implement was merging most or all of rustc_llvm into rustc_trans. I ran into a nasty bug, which was probably just a silly typo somewhere but I probably won't have the time to figure it out in the next week or two. I opened #41699 for that step.

Fixes #41473

Commit:
Remove rustc_llvm dependency from rustc_metadata

Move the code for loading metadata from rlibs and dylibs from
rustc_metadata into rustc_trans, and introduce a trait to avoid
introducing a direct dependency on rustc_trans.

This means rustc_metadata is no longer rebuilt when LLVM changes.

rust-highfive commented:
r? @arielb1

(rust_highfive has picked a reviewer for you, use r? to override)

alexcrichton commented:
Thanks for the PR! We’ll periodically check in on it to make sure that @arielb1 or someone else from the team reviews it soon.

Commit:
Remove rustc_llvm dependency from librustc

Consequently, session creation can no longer initialize LLVM.
The few places that use the compiler without going through
rustc_driver/CompilerCalls thus need to be careful to manually
initialize LLVM (via rustc_trans!) immediately after session
creation.

This means librustc is not rebuilt when LLVM changes.

Commit:
Remove (direct) rustc_llvm dependency from rustc_driver

This does not actually improve build times, since it still depends
on rustc_trans, but is better layering and fits the multi-backend
future slightly better.

bors commented:
:umbrella: The latest upstream changes (presumably #41560) made this pull request unmergeable. Please resolve the merge conflicts.

rkruppe commented:
This is now ready for review, though it still can't merge because of #27812. I updated the PR description to summarize what this PR does and doesn't do.

rkruppe commented:
Fixed the travis failure, I missed one place in rustdoc (as I had feared when I did this refactoring 😅 ).

Mark-Simulacrum commented:
Looks like there are a few more test failures:

```
[00:53:00] failures:
[00:53:00]     [run-make] run-make/issue-19371
[00:53:00]     [run-make] run-make/sysroot-crates-are-unstable
```

rkruppe commented:
Fixed `run-make/issue-19371`. The other failure is expected as https://github.com/rust-lang/rust/issues/27812 is still open. Fixing it involves either adding `#![cfg_attr(rustbuild, unstable(...))]` to the crates.io packages pulled in (boo) or @eddyb implementing `-Z force-unstable=rustc_private`. Either way, this PR can be reviewed independently of that.

arielb1 commented on src/librustc_trans/lib.rs:
```diff
@@ -180,3 +180,55 @@ pub struct CrateTranslation {
 __build_diagnostic_array! { librustc_trans, DIAGNOSTICS }
+
+use rustc::session::Session;
+pub fn init(sess: &Session) {
```
Could you make this return a token so that you can't call the 10 or so public trans functions without calling `init` (AFAICT half of these - `provide`, `find_crate_name`, `filename_for_input`, `default_output_for_target`, `invalid_output_for_target` - don't need LLVM, but a few others do)?

arielb1 reviewed (approved):
nits are minor. fix these if you want.

r+ modulo `owning_ref` working

arielb1 commented on src/librustc_trans/llvm_util.rs:
```diff
@@ -0,0 +1,120 @@
+use libc::{c_int, c_char};
+use std::ffi::CString;
+
+pub fn init(sess: &Session) {
```
Could you make this return a token so that you can't call the 10 or so public trans functions without calling init (AFAICT half of these - `provide`, `find_crate_name`, `filename_for_input`, `default_output_for_target`, `invalid_output_for_target` - don't need LLVM, but a few others do)?

eddyb replied:
Can't they just call `init` themselves? The `Once` would prevent misuse.

rkruppe replied:
I'm a bit torn on this. Threading a token everywhere is a lot of infrastructure for the benefit IMHO. Initializing when calling into trans solves that, but needs care when exposing new functions from trans. Requiring the driver to call init obviously places a burden on the driver, but it's a relatively simple one (a single call, and can be the same for all backends) that needs to be observed in fewer places.

rkruppe replied:
I've decided to punt on this for the time being. I hope we'll get something like a `Backend` trait object, which would solve this issue without having to audit every public API of trans.

bors commented:
:umbrella: The latest upstream changes (presumably #41818) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented on src/librustc/middle/cstore.rs:
```diff
@@ -268,6 +286,9 @@ pub trait CrateStore {
     fn metadata_encoding_version(&self) -> &[u8];
+
+    // access to the metadata loader
+    fn metadata_loader(&self) -> &MetadataLoader;
```
Can you move this next to `crate_data_as_rc_any` (at the top of the trait)? They're both kind of special.

eddyb commented on src/librustc_driver/lib.rs:
```diff
@@ -652,12 +657,10 @@ impl RustcDefaultCalls {
                 PrintRequest::TargetCPUs => {
-                    let tm = create_target_machine(sess);
-                    unsafe { llvm::LLVMRustPrintTargetCPUs(tm); }
+                    rustc_trans::print_target_cpus(sess);
```
Can this be a single entry-point that takes a `PrintRequest`?

rkruppe replied:
Yes, I just didn't realize how when I wrote this code. One advantage of leaving a PR to rot for a week: You get a fresh perspective :smile:

eddyb commented on src/librustc_driver/lib.rs:
```diff
@@ -710,10 +713,7 @@ pub fn version(binary: &str, matches: &getopts::Matches) {
-            println!("LLVM version: {}.{}",
-                     llvm::LLVMRustVersionMajor(), llvm::LLVMRustVersionMinor());
-        }
+        rustc_trans::print_llvm_version();
```
Can this be `print_version`? Other backends would print other information.

rkruppe replied:
Will do.

eddyb commented on src/librustc_driver/test.rs:
```diff
@@ -103,13 +104,14 @@ fn test_env<F>(source_string: &str,
     let dep_graph = DepGraph::new(false);
     let _ignore = dep_graph.in_ignore();
-    let cstore = Rc::new(CStore::new(&dep_graph));
+    let cstore = Rc::new(CStore::new(&dep_graph, box rustc_trans::LlvmMetadataLoader));
```
Can this not use a dummy `CStore`?

eddyb commented on src/librustc_trans/llvm_util.rs:
```diff
@@ -0,0 +1,120 @@
+        // Before we touch LLVM, make sure that multithreading is enabled.
+        use std::sync::Once;
+        static INIT: Once = Once::new();
+        static mut POISONED: bool = false;
```
I'd be way more comfortable with an `AtomicBool` here.

rkruppe replied:
The thought also crossed my mind as I cut and pasted the code, but I erred on the side of only moving code. I'll change it.

nikomatsakis commented on src/librustc/Cargo.toml:
```diff
@@ -13,12 +13,12 @@ arena = { path = "../libarena" }
 fmt_macros = { path = "../libfmt_macros" }
 graphviz = { path = "../libgraphviz" }
 log = "0.3"
+owning_ref = "0.3.3"
```
So @eddyb pointed this out to me -- I'm not sure if we should add a dependency on this. First off i'm not sure if there are technical implications, but also, it's not under the same license as rust (just MIT, not MIT/Apache2).

cc @brson @alexcrichton

eddyb replied:
The reason I mentioned it, fwiw, was that we might have more uses for it.
OTOH, my initial suggestion for this PR was `Box<Deref<Target=[u8]>>`, so `owning_ref` is *potentially* overkill (although I believe it is more efficient).

rkruppe replied:
I can't imagine the license of owning_ref is any more of a problem than, say, LLVM or jemalloc being under a different license (jemalloc is even linked into user code, not just rustc) . But of course IANAL and this is a valid question to ask.

FWIW it wouldn't be hard to write something equivalent to `ErasedBoxRef`, especially if it doesn't need a safe general purpose interface. But, needless to say, it would be quite regrettable if we kept duplicating crates.io packages.

nikomatsakis replied:
@rkruppe your argument re: LLVM etc makes sense to me. I'm curious to hear what @brson has to say since I consider him more the expert on this stuff than me. =)

(In general, though, I'm excited about gaining the ability to use external crates in the compiler.)

nikomatsakis replied:
OK, @alexcrichton tells me that we have already whitelisted this, so it's all good!

brson replied:
This is fine to pull in owning_ref under MIT. We have historically accepted minor dependencies under permissive licenses including MIT and BSD, as these are maximally compatible with Rust's other licenses.

There's a tidy check that tries to keep our licenses in order but it may have some gaps yet.

Thanks for pinging me @nikomatsakis

Commit:
Use AtomicBool instead of a 'static mut' for LLVM init posioning

rkruppe commented:
Rebased and addressed all nits except the LLVM init token / make-trans-init-itself thing.

eddyb commented:
@rkruppe Please check Travis failure.

rkruppe commented:
Ugh, I've fixed that before and somehow lost that in a rebase. Fixed again.

arielb1 commented:
@bors r+

bors commented:
:pushpin: Commit acdeedc has been approved by `arielb1`

<!-- @bors r=arielb1 acdeedc19253130798916a8bb20651657d9f01e5 -->

bors commented:
:hourglass: Testing commit acdeedc with merge b8742f4...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/232386234?utm_source=github_status&utm_medium=notification)

Mark-Simulacrum commented:
Legitimate failure, presumably: `[01:22:46] thread '[run-make] run-make/llvm-pass' panicked at 'explicit panic', src/tools/compiletest/src/runtest.rs:2472`.

Commit:
Fix run-make/llvm-pass

rkruppe commented:
Oops! I thought I'd tested run-make. Maybe it was on Windows, where that test is disabled? Anyway, fixed.

rkruppe commented:
(Well, *hopefully* fixed, I can't test right now due to being on Windows.)

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit 04a16ff has been approved by `eddyb`

<!-- @bors r=eddyb 04a16ff5acaaccf6332f706f090442f0ab9b0f95 -->

bors commented:
:hourglass: Testing commit 04a16ff with merge ceef6c9...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3263)

alexcrichton commented:
@bors: retry

* network failure https://github.com/rust-lang/rust/issues/40474

bors commented:
:hourglass: Testing commit 04a16ff with merge ed1bd1a...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3265)

Mark-Simulacrum commented:
@bors retry
 - Handle.exe network failure #40474.

bors commented:
:hourglass: Testing commit 04a16ff with merge 8ec9946...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3267)

alexcrichton commented:
@bors: retry 

* network issues https://github.com/rust-lang/rust/issues/40474

bors commented:
:hourglass: Testing commit 04a16ff with merge fa78d5b...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3275), [status-travis](https://travis-ci.org/rust-lang/rust/builds/232694345?utm_source=github_status&utm_medium=notification)
Approved by: eddyb
Pushing fa78d5bb87d136f793c56f9861957ce41080bd7e to master...
